### PR TITLE
Refactor pointers planner to smart pointers

### LINKED ISSUE #99: CleanPlan is not working properly
```
create table test1(a int);
create table test2(a int);
insert into test1 values (1);
insert into test2 values (2);

select test1.a + 1 from test1,test2;
```

If this sequence of queries are executed, the peloton fails with segmentation fault.
Our team looked into the stack trace and found that the following line causes the error.
If we comment out the "delete root", it no longer causes segmentation but it may possibly lead
the memory leak. 

https://github.com/cmu-db/peloton/blob/master/src/backend/bridge/dml/mapper/mapper.cpp#L195

I include the stack trace.
```
#0  0x0000000000000085 in ?? ()
#1  0x00007ffff6d7d776 in std::default_delete<peloton::catalog::Schema const>::operator() (this=0x7fffdc0bef10, __ptr=0x7fffdc0bee60) at /usr/include/c++/4.8/bits/unique_ptr.h:67
#2  0x00007ffff6d7d511 in std::unique_ptr<peloton::catalog::Schema const, std::default_delete<peloton::catalog::Schema const> >::~unique_ptr (this=0x7fffdc0bef10, __in_chrg=<optimized out>) at /usr/include/c++/4.8/bits/unique_ptr.h:184
#3  0x00007ffff6d7d816 in peloton::planner::ProjectionPlan::~ProjectionPlan (this=0x7fffdc0beee0, __in_chrg=<optimized out>) at ../../src/backend/planner/projection_plan.h:29
#4  0x00007ffff6d7d862 in peloton::planner::ProjectionPlan::~ProjectionPlan (this=0x7fffdc0beee0, __in_chrg=<optimized out>) at ../../src/backend/planner/projection_plan.h:29
#5  0x00007ffff6d79b65 in peloton::bridge::PlanTransformer::CleanPlan (root=0x7fffdc0beee0) at ../../src/backend/bridge/dml/mapper/mapper.cpp:194
#6  0x00007ffff6d7b7eb in std::_Sp_counted_deleter<peloton::planner::AbstractPlan const*, void (*)(peloton::planner::AbstractPlan const*), std::allocator<int>, (__gnu_cxx::_Lock_policy)2>::_M_dispose (this=0x7fffdc0d3870) at /usr/include/c++/4.8/bits/shared_ptr_base.h:347
#7  0x00007ffff7779882 in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x7fffdc0d3870) at /usr/include/c++/4.8/bits/shared_ptr_base.h:144
#8  0x00007ffff777960f in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count (this=0x7fffeaae0bf8, __in_chrg=<optimized out>) at /usr/include/c++/4.8/bits/shared_ptr_base.h:546
#9  0x00007ffff77795b2 in std::__shared_ptr<peloton::planner::AbstractPlan const, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr (this=0x7fffeaae0bf0, __in_chrg=<optimized out>) at /usr/include/c++/4.8/bits/shared_ptr_base.h:781
#10 0x00007ffff77795cc in std::shared_ptr<peloton::planner::AbstractPlan const>::~shared_ptr (this=0x7fffeaae0bf0, __in_chrg=<optimized out>) at /usr/include/c++/4.8/bits/shared_ptr.h:93
#11 0x00007ffff7778f73 in peloton_dml (planstate=0x7fffdc0e7670, sendTuples=true, dest=0x7fffdc0b89a0, tuple_desc=0x7fffdc0e97f8, prepStmtName=0x0) at ../../src/postgres/backend/postmaster/peloton.cpp:201
#12 0x00007ffff7620733 in peloton_ExecutePlan (estate=0x7fffdc0e7560, planstate=0x7fffdc0e7670, operation=CMD_SELECT, sendTuples=true, numberTuples=0, direction=ForwardScanDirection, dest=0x7fffdc0b89a0, tupDesc=0x7fffdc0e97f8, prepStmtName=0x0) at ../../src/postgres/backend/executor/execMain.cpp:1658
#13 0x00007ffff761e863 in standard_ExecutorRun (queryDesc=0x7fffdc0be910, direction=ForwardScanDirection, count=0) at ../../src/postgres/backend/executor/execMain.cpp:378
#14 0x00007ffff761e6d1 in ExecutorRun (queryDesc=0x7fffdc0be910, direction=ForwardScanDirection, count=0) at ../../src/postgres/backend/executor/execMain.cpp:300
#15 0x00007ffff7808e2b in PortalRunSelect (portal=0x7fffdc0b68d0, forward=true, count=0, dest=0x7fffdc0b89a0) at ../../src/postgres/backend/tcop/pquery.cpp:860
#16 0x00007ffff7808a7c in PortalRun (portal=0x7fffdc0b68d0, count=9223372036854775807, isTopLevel=true, dest=0x7fffdc0b89a0, altdest=0x7fffdc0b89a0, completionTag=0x7fffeaae0fe0 "") at ../../src/postgres/backend/tcop/pquery.cpp:714
#17 0x00007ffff78021e4 in exec_simple_query (query_string=0x7fffdc068130 "select test1.a + 1 from test1,test2;") at ../../src/postgres/backend/tcop/postgres.cpp:1118
#18 0x00007ffff7806b3a in PostgresMain (argc=1, argv=0x7fffdc001f88, dbname=0x7fffdc000c78 "postgres", username=0x7fffdc000c10 "vagrant") at ../../src/postgres/backend/tcop/postgres.cpp:4054
#19 0x00007ffff778bd10 in BackendRun (port=0x60a620) at ../../src/postgres/backend/postmaster/postmaster.cpp:4015
#20 0x00007ffff778b0cb in BackendTask (bn=0x60a7e0, port=0x60a620, param=0x60db60) at ../../src/postgres/backend/postmaster/postmaster.cpp:3634
#21 0x00007ffff7790118 in std::_Bind_simple<void (*(bkend*, Port*, BackendParameters*))(bkend*, Port*, BackendParameters*)>::_M_invoke<0ul, 1ul, 2ul>(std::_Index_tuple<0ul, 1ul, 2ul>) (this=0x6063f0) at /usr/include/c++/4.8/functional:1732
#22 0x00007ffff778ff39 in std::_Bind_simple<void (*(bkend*, Port*, BackendParameters*))(bkend*, Port*, BackendParameters*)>::operator()() (this=0x6063f0) at /usr/include/c++/4.8/functional:1720
#23 0x00007ffff778fe88 in std::thread::_Impl<std::_Bind_simple<void (*(bkend*, Port*, BackendParameters*))(bkend*, Port*, BackendParameters*)> >::_M_run() (this=0x6063d8) at /usr/include/c++/4.8/thread:115
#24 0x00007ffff56a0a60 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#25 0x00007ffff4c79182 in start_thread (arg=0x7fffeaae8700) at pthread_create.c:312
#26 0x00007ffff632f47d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```

### LINKED ISSUE #120: Fix Copy() interface in Expression subsystem to return unique_ptr instead of raw pointers.
We need a similar refactoring of the Copy() interface as was done in #118 for AbstractExpression. @amaliujia can you take care of this?

### PR DESCRIPTION
**Pull request by omegaga**
This PR is intended to fix #99. Thanks to comments by @pmenon, I now use unique_ptr in all planner node. In this version, I still use shared_ptr for a plan node in plan cache, and for schema in projection, join and aggregate plans.

Remind that expression, executor and many other places still use raw pointers. So there're still (ugly) conversion from smart pointer to raw pointer now. Hopefully we can change all those code and eliminate raw pointers.

This PR also fixes #120.

This PR changes a large number of files. Please review and leave comments if you find something ridiculous :-)

### CODE REVIEW

**Comment by amaliujia** [comment]
For plan children, I am curious why change unique_ptr to share_ptr? Do you plan to share the plan nodes themselves?

**Comment by omegaga** [comment]
@amaliujia Children are stored as unique_ptr (https://github.com/omegaga/peloton/blob/fixbug/src/backend/planner/abstract_plan.h#L82).

The reason is that we want to ensure each node is used exactly once in the plan tree. You can see #102 for discussion.

In rare cases when you need to take a subtree out, it's easy to transfer its ownership to a shared_ptr.

**Comment by amaliujia** [comment]
O, sorry. I mixed the changed part with unchanged part.

**Comment by pmenon** [comment]
I took a quick glance at this and noticed that the getters in the plan nodes are returning `unique_ptr`.  They should be returning raw pointers if the value can be null, or references if they're guaranteed not to be null.  Remember, smart pointers are used to reflect **ownership**.  By returning a `unique_ptr`, the interface is saying that the caller is **taking ownership** of the object that is being returned.  If the caller is not taking ownership in any way, return a reference (or a regular pointer if the value can be null).

File: src/backend/planner/abstract_join_plan.h [hunk 1]
```diff
@@ -54,13 +55,18 @@ class AbstractJoinPlan : public AbstractPlan {
 
   PelotonJoinType GetJoinType() const { return join_type_; }
 
-  const expression::AbstractExpression *GetPredicate() const {
-    return predicate_.get();
+  const std::unique_ptr<const expression::AbstractExpression> &GetPredicate()
+      const {
+    return predicate_;
```
Comment (pmenon) [inline src/backend/planner/abstract_join_plan.h]: This shouldn't be returning unique_ptr

File: src/backend/planner/abstract_join_plan.h [hunk 2]
```diff
@@ -54,13 +55,18 @@ class AbstractJoinPlan : public AbstractPlan {
...
 
-  const ProjectInfo *GetProjInfo() const { return proj_info_.get(); }
+  const std::unique_ptr<const ProjectInfo> &GetProjInfo() const {
+    return proj_info_;
+  }
 
-  const catalog::Schema *GetSchema() const { return proj_schema_.get(); }
+  const std::shared_ptr<const catalog::Schema> &GetSchema() const {
+    return proj_schema_;
+  }
```
Comment (pmenon) [inline src/backend/planner/abstract_join_plan.h]: This shouldn't be returning unique_ptr

File: src/backend/planner/abstract_join_plan.h [hunk 3]
```diff
@@ -54,13 +55,18 @@ class AbstractJoinPlan : public AbstractPlan {
...
-    return predicate_.get();
+  const std::unique_ptr<const expression::AbstractExpression> &GetPredicate()
+      const {
+    return predicate_;
   }
 
-  const ProjectInfo *GetProjInfo() const { return proj_info_.get(); }
+  const std::unique_ptr<const ProjectInfo> &GetProjInfo() const {
+    return proj_info_;
+  }
```
Comment (pmenon) [inline src/backend/planner/abstract_join_plan.h]: This shouldn't be returning unique_ptr

File: src/backend/planner/aggregate_plan.h [hunk 4]
```diff
@@ -63,20 +64,21 @@ class AggregatePlan : public AbstractPlan {
     return groupby_col_ids_;
   }
 
-  const expression::AbstractExpression *GetPredicate() const {
-    return predicate_.get();
+  const std::unique_ptr<const expression::AbstractExpression> &GetPredicate()
+      const {
+    return predicate_;
```
Comment (pmenon) [inline src/backend/planner/aggregate_plan.h]: This shouldn't be returning unique_ptr

File: src/backend/planner/aggregate_plan.h [hunk 5]
```diff
@@ -63,20 +64,21 @@ class AggregatePlan : public AbstractPlan {
...
   }
 
-  const planner::ProjectInfo *GetProjectInfo() const {
-    return project_info_.get();
+  const std::unique_ptr<const planner::ProjectInfo> &GetProjectInfo() const {
+    return project_info_;
```
Comment (pmenon) [inline src/backend/planner/aggregate_plan.h]: This shouldn't be returning unique_ptr

**Comment by omegaga** [comment]
@pmenon I returned a constant reference in the functions you mentioned above, so the ownership is not transferred in this case. BTW, why do I need to use a raw pointer if it can be a nullptr? I think unique_ptr can also wrap a nullptr.

**Comment by pmenon** [comment]
The problem is that you're exposing how the object is internally managed to external clients.  Think about the case where you had to change the schema to use a shared_ptr.  With your code, you'd have to change the interface **and** and users of the getter. Returning a reference requires no interface changes, nor client changes.

**Comment by omegaga** [comment]
Hmm, this makes sense. I will change it back to raw pointers.

**Comment by omegaga** [comment]
Push a new commit. Now getter methods return constant raw pointers

**Pull request merged**